enterprise/server/remote_execution/copy_on_write/copy_on_write.go

```diff
@@ -615,51 +622,45 @@ func (s *COWStore) initDirtyChunk(offset int64, size int64) (ogChunk *Mmap, newC
 	return ogChunk, newChunk, nil
 }
 
-type eagerFetchData struct {
-	// Chunk offset to eagerly fetch
-	offset int64
-}
-
 func (s *COWStore) eagerFetchNextChunks(offset int64) {
 	currentOffset := offset + s.chunkSizeBytes
 	for i := 0; i < numChunksToEagerFetch; i++ {
-		s.sendNonBlockingEagerFetch(currentOffset)
+		s.eagerFetchStack.Push(currentOffset)
 		currentOffset += s.chunkSizeBytes
 	}
 }
 
-func (s *COWStore) sendNonBlockingEagerFetch(offset int64) {
-	select {
-	case s.eagerFetchChan <- &eagerFetchData{offset: offset}:
-	default:
-	}
-}
-
 func (s *COWStore) eagerFetchChunksInBackground() {
 	rateLimiter := rate.NewLimiter(rate.Limit(*maxEagerFetchesPerSec), 1)
 	eg := &errgroup.Group{}
 	eg.SetLimit(*eagerFetchConcurrency)
 	defer eg.Wait()
 
+	ctx, cancel := context.WithCancel(s.ctx)
+	defer cancel()
+	go func() {
+		<-s.quitChan
+		cancel()
+	}()
+
 	for {
-		select {
-		case <-s.quitChan:
-			return
-		case d := <-s.eagerFetchChan:
-			if err := rateLimiter.Wait(s.ctx); err != nil {
-				if err != s.ctx.Err() {
-					log.CtxErrorf(s.ctx, "COWStore eager fetch rate limiter failed, stopping eager fetches: %s", err)
-				}
-				return
+		offset, err := s.eagerFetchStack.Recv(ctx)
+		if err != nil {
+			break // context canceled
+		}
+		if err := rateLimiter.Wait(ctx); err != nil {
+			if err != s.ctx.Err() {
+				log.CtxErrorf(s.ctx, "COWStore eager fetch rate limiter failed, stopping eager fetches: %s", err)
 			}
-			eg.Go(func() error {
-				err := s.fetchChunk(d.offset)
-				if err != nil {
-					log.CtxWarningf(s.ctx, "COWStore eager fetch chunk failed with: %s", err)
-				}
-				return nil
-			})
+			return
 		}
+		eg.Go(func() error {
+			err := s.fetchChunk(offset)
+			if err != nil {
```

Comment:
can scope this err down:

if err := s.fetchChunk(offset); err != nil {
 log da err
}

Reply:
done